### PR TITLE
ISSUE-325 Modify APIs to support rendering UI for schema branch

### PR DESCRIPTION
**Pull request by raju-saravanan**
ISSUE-325 Modify APIs to support rendering UI for schema branch, #325

### CODE REVIEW

**Comment by raju-saravanan** [comment]
@satishd , this PR will break the existing UI,  maybe we should hold this PR until @shahsank3t pushes in his UI changes?

**Comment by raju-saravanan** [comment]
@shahsank3t : These are the list of API exposed as part of "branching" feature

-  **POST /schemas/versionsById/{versionId}/branch**
    This is to create a new branch from an existing Schema Version ID from MASTER
- **POST /schemas/s2/versions?branch={branchId}**
    This is to add a new schema version to a branch, this is an existing API which takes another query 
    param branchId, if the branchId is not mentioned then it assumes that the version has to be 
    added to 'MASTER' 
- **POST /schemas/{versionId}/merge**
    This is to merge a schemaVersion from a branch to 'MASTER'
- **POST /schemas/branch/{branchId}**
    This is to delete a schema branch given its branchId

I have changed the structure of the AggregatedSchemaMetadata, which include schema branch info  as well, you can hit it and let me know if you have any queries on the JSON structure

**Comment by satishd** [comment]
@raju-saravanan Can you put the REST APIs that were changed so that @shahsank3t can make respective changes in UI? You may want to mention about the removed field/s and new field/s added.

**Comment by raju-saravanan** [comment]
@satishd I will make a document around the APIs and share the link here.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/AggregatedSchemaBranch.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.io.Serializable;
+import java.util.Collection;
+
+/*
+     This class represents the aggregated information about the schema branch which includes all the schema versions tied
+     to that branch and the schema version from which the branch was created
+ */
```
Comment (satishd) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/AggregatedSchemaBranch.java]: nit: Follow class level javadoc style like we do in other classes.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/AggregatedSchemaBranch.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AggregatedSchemaBranch implements Serializable {
+
+    private static final long serialVersionUID = -4013250198049024761L;
+
+    private SchemaBranch schemaBranch;
+    private Long rootSchemaVersion;
+    private Collection<SchemaVersionInfo> schemaVersionInfos;
+
+    public AggregatedSchemaBranch(SchemaBranch schemaBranch, Long rootSchemaVersion, Collection<SchemaVersionInfo> schemaVersionInfos) {
```
Comment (satishd) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/AggregatedSchemaBranch.java]: This should have no-arg default constructor for creating these objects from json text. Same needs to be also fixed in `AggregatedSchemaMetadataInfo`. 

Pl add a test to check that. I tried below to check that and it gets failed. 

```
        String aggregateSchemaMetadataStr = new ObjectMapper().writeValueAsString(aggregatedSchemaMetadataCollection);
        LOG.info("aggregateSchemaMetadataStr = " + aggregateSchemaMetadataStr);

        Collection<AggregatedSchemaMetadataInfo> returnedResult =
                new ObjectMapper().readValue(aggregateSchemaMetadataStr,
                                             new TypeReference<Collection<AggregatedSchemaMetadataInfo>>() {});

        LOG.info("returnedResult = " + returnedResult);
        for (AggregatedSchemaMetadataInfo aggregatedSchemaMetadataInfo : returnedResult) {
            LOG.info("aggregatedSchemaMetadataInfo = " + aggregatedSchemaMetadataInfo);
        }
```

File: schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java [hunk 3]
```diff
@@ -538,6 +529,16 @@ public SchemaIdVersion mergeSchemaVersion(Long schemaVersionId, SchemaVersionMer
         return schemaVersionLifecycleManager.mergeSchemaVersion(schemaVersionId, schemaVersionMergeStrategy);
     }
 
+    @Override
+    public Collection<AggregatedSchemaBranch> getAggregatedSchemaBranch(String schemaName) throws SchemaNotFoundException, SchemaBranchNotFoundException {
+        Collection<AggregatedSchemaBranch> aggregatedSchemaBranches = new LinkedList<>();
```
Comment (satishd) [inline schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java]: nit: may want to use arraylist to avoid creating more references with linkedlist, there may not be  large no of branches for a schema.

File: schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java [hunk 4]
```diff
@@ -561,7 +562,13 @@ public SchemaBranch createSchemaBranch(Long schemaVersionId, SchemaBranch schema
 
         Preconditions.checkNotNull(schemaBranch.getName(), "Schema branch name can't be null");
 
-        SchemaBranch existingSchemaBranch = schemaBranchCache.getIfPresent(SchemaBranchCache.Key.of(schemaBranch.getName()));
+        SchemaBranch existingSchemaBranch = null;
+        try {
+            existingSchemaBranch = schemaBranchCache.get(SchemaBranchCache.Key.of(schemaBranch.getName()));
```
Comment (satishd) [inline schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java]: Why are we changing from `getIfPresent` to `get` and catching the exception?

File: schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/SchemaVersionLifecycleManager.java [hunk 5]
```diff
@@ -444,7 +444,8 @@ private CompatibilityResult checkCompatibility(String type,
         Collection<SchemaVersionInfo> schemaVersionInfos;
 
         if(schemaBranchName.equals(SchemaBranch.MASTER_BRANCH)) {
-            List<QueryParam> queryParams = Collections.singletonList(new QueryParam(SchemaVersionStorable.NAME, schemaName));
+            List<QueryParam> queryParams = Collections.singletonList(new QueryParam(SchemaVersionStorable.SCHEMA_METADATA_ID,
```
Comment (satishd) [inline schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/SchemaVersionLifecycleManager.java]: We need to change param name in the existing code from `SchemaVersionStorable.NAME` to `SchemaMetadataStorable.NAME` to be consistent.
Why is this changed to use metadata-id?

**Comment by raju-saravanan** [comment]
@satishd: Addressed all the PR comments.

**Comment by shahsank3t** [comment]
@satishd 
I have asked @satishwaghela to work on the UI changes. If it's ok, he will be creating his own PR by today EOD, so I can review once before you guys can merge it in.
